src/core/server/deprecations/deprecations_factory.ts

```diff
@@ -0,0 +1,112 @@
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { DeprecationsRegistry } from './deprecations_registry';
+import type { Logger } from '../logging';
+import type {
+  PluginDeprecationDetails,
+  DeprecationsDetails,
+  GetDeprecationsContext,
+} from './types';
+
+export interface DeprecationsFactoryDeps {
+  logger: Logger;
+}
+
+export class DeprecationsFactory {
+  private readonly registries: Map<string, DeprecationsRegistry> = new Map();
+  private readonly logger: Logger;
+  constructor({ logger }: DeprecationsFactoryDeps) {
+    this.logger = logger;
+  }
+
+  public createRegistry = (pluginId: string): DeprecationsRegistry => {
+    const exsiting = this.registries.get(pluginId);
+    if (exsiting) {
+      return exsiting;
+    }
+    const registry = new DeprecationsRegistry();
+    this.registries.set(pluginId, registry);
+    return registry;
+  };
+
+  public getRegistry = (pluginId: string): DeprecationsRegistry | undefined => {
+    return this.registries.get(pluginId);
+  };
+
+  public getDeprecations = async (
+    pluginId: string,
+    dependencies: GetDeprecationsContext
+  ): Promise<PluginDeprecationDetails[]> => {
+    const infoBody = await this.getDeprecationsBody(pluginId, dependencies);
+    return this.createDeprecationInfo(pluginId, infoBody).flat();
+  };
+
+  public getAllDeprecations = async (
+    dependencies: GetDeprecationsContext
+  ): Promise<PluginDeprecationDetails[]> => {
+    const pluginIds = [...this.registries.keys()];
+
+    const deprecationsInfo = await Promise.all(
+      pluginIds.map(async (pluginId) => {
+        const infoBody = await this.getDeprecationsBody(pluginId, dependencies);
+        return this.createDeprecationInfo(pluginId, infoBody);
+      })
+    );
+
+    return deprecationsInfo.flat();
+  };
+
+  private createDeprecationInfo = (
+    pluginId: string,
+    deprecationInfoBody: DeprecationsDetails[]
+  ): PluginDeprecationDetails[] => {
+    return deprecationInfoBody
+      .flat()
+      .filter(Boolean)
+      .map((pluginDeprecation) => ({
+        ...pluginDeprecation,
+        pluginId,
+      }));
+  };
+
+  private getDeprecationsBody = async (
+    pluginId: string,
+    dependencies: GetDeprecationsContext
+  ): Promise<DeprecationsDetails[]> => {
+    const deprecationsRegistry = this.registries.get(pluginId);
+    if (!deprecationsRegistry) {
+      return [];
+    }
+    try {
+      const settledResults = await deprecationsRegistry.getDeprecations(dependencies);
+      return settledResults.flatMap((settledResult) => {
+        if (settledResult.status === 'rejected') {
+          this.logger.warn(
+            `Failed to get deprecations info for plugin "${pluginId}".`,
+            settledResult.reason
+          );
+          return [
+            {
+              message: `Failed to get deprecations info for plugin "${pluginId}".`,
+              level: 'warning',
```

Comment:
Should we distinguish the rejected state with a different level? Or something else? If we end up in this state, we might want to display it differently in the UI (with a callout or some sort of error notification). It would be nice if we had an easy way to identify it other than parsing the message string.

Reply:
Do you think it is enough to identify it with `level: 'fetch_error'` or something like that? `pluginId` is already provided in the payload as well.

Reply:
👍 I think that would work.